### PR TITLE
[CH-76]Union support columnar

### PR DESCRIPTION
**Pull request by lgbo-ustc**
## What changes were proposed in this pull request?

add support for running union operation on native engine

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

### CODE REVIEW

**Review by rui-mo — COMMENTED** [review]
Generally LGTM. Will test TPC-DS on Velox backend.

File: jvm/src/main/scala/io/glutenproject/execution/BasicPhysicalOperatorTransformer.scala [hunk 1]
```diff
@@ -34,7 +31,7 @@ import io.glutenproject.substrait.plan.PlanBuilder
 import io.glutenproject.substrait.rel.{RelBuilder, RelNode}
 import io.glutenproject.utils.BindReferencesUtil
 import io.glutenproject.vectorized.{ExpressionEvaluator, OperatorMetrics}
-
+import io.substrait.proto.SetRel
```
Comment (rui-mo) [inline jvm/src/main/scala/io/glutenproject/execution/BasicPhysicalOperatorTransformer.scala:34]: Is this still in use?

File: jvm/src/main/scala/io/glutenproject/execution/BasicPhysicalOperatorTransformer.scala [hunk 2]
```diff
@@ -507,10 +543,41 @@ case class UnionExecTransformer(children: Seq[SparkPlan]) extends SparkPlan with
...
+      true
+    }
+  }
+
 
   override def doTransform(context: SubstraitContext): TransformContext = {
-    throw new UnsupportedOperationException(s"This operator doesn't support doTransform.")
+    val operatorId = context.nextOperatorId
+    val attrList = new util.ArrayList[Attribute]()
+    for (attr <- children(0).output)
```
Comment (rui-mo) [inline jvm/src/main/scala/io/glutenproject/execution/BasicPhysicalOperatorTransformer.scala:574]: How about adding some check for whether children is empty here?

File: jvm/src/main/scala/io/glutenproject/execution/BasicPhysicalOperatorTransformer.scala [hunk 3]
```diff
@@ -492,7 +516,19 @@ case class UnionExecTransformer(children: Seq[SparkPlan]) extends SparkPlan with
   }
 
   override def columnarInputRDDs: Seq[RDD[ColumnarBatch]] = {
-    throw new UnsupportedOperationException(s"This operator doesn't support inputRDDs.")
+    val retRDD = children.map {
+      case c: TransformSupport =>
+        val r: Seq[RDD[ColumnarBatch]] = c.columnarInputRDDs
```
Comment (zzcclp) [inline jvm/src/main/scala/io/glutenproject/execution/BasicPhysicalOperatorTransformer.scala:521]: directly use `c.columnarInputRDDs`

File: jvm/src/main/scala/io/glutenproject/execution/BasicPhysicalOperatorTransformer.scala [hunk 4]
```diff
@@ -492,7 +516,19 @@ case class UnionExecTransformer(children: Seq[SparkPlan]) extends SparkPlan with
...
+        r
+      case c =>
+        val r = Seq(c.executeColumnar())
```
Comment (zzcclp) [inline jvm/src/main/scala/io/glutenproject/execution/BasicPhysicalOperatorTransformer.scala:524]: directly use `Seq(c.executeColumnar())`

File: backends-clickhouse/src/test/scala/io/glutenproject/execution/GlutenClickHouseTPCDSParquetSuite.scala [hunk 5]
```diff
@@ -83,6 +83,19 @@ class GlutenClickHouseTPCDSParquetSuite extends GlutenClickHouseTPCDSAbstractSui
     assert(result(0).getDouble(1) == 80037.12727449503)
   }
 
+  test("test union operator") {
```
Comment (zzcclp) [inline backends-clickhouse/src/test/scala/io/glutenproject/execution/GlutenClickHouseTPCDSParquetSuite.scala:86]: maybe add more test cases, for example, three union.

File: jvm/src/main/java/io/glutenproject/substrait/rel/RelBuilder.java [hunk 6]
```diff
@@ -27,7 +27,6 @@
 import io.substrait.proto.JoinRel;
 import io.substrait.proto.SortField;
 import org.apache.spark.sql.catalyst.expressions.Attribute;
-
```
Comment (zzcclp) [inline jvm/src/main/java/io/glutenproject/substrait/rel/RelBuilder.java:30]: fallback

File: jvm/src/main/scala/io/glutenproject/execution/BasicPhysicalOperatorTransformer.scala (hunk 2 above)
Comment (lgbo-ustc) [reply jvm/src/main/scala/io/glutenproject/execution/BasicPhysicalOperatorTransformer.scala:575]: add check in `doValidate()`

**Comment by zzcclp** [comment]
please rebase to main.

@rui-mo do you verified TPC-DS and TPCH on Velox Backend?

**Comment by rui-mo** [comment]
> please rebase to main.
> 
> @rui-mo do you verified TPC-DS and TPCH on Velox Backend?

Will do that today, thanks.

**Comment by rui-mo** [comment]
Tested this PR on TPC-DS, but found union fallbacked due to the input iter cannot be found. Will check what support is needed for Velox backend.

File: backends-velox/src/test/scala/io/glutenproject/execution/TestOperator.scala [hunk 7]
```diff
@@ -184,42 +184,4 @@ class TestOperator extends WholeStageTransformerSuite {
       Seq(Row(48.0), Row(10.0), Row(23.0), Row(38.0), Row(75.0), Row(33.0), Row(93.0))
     TestUtils.compareAnswers(result, expected)
   }
-
-  test("test_union_all two tables") {
```
Comment (rui-mo) [inline backends-velox/src/test/scala/io/glutenproject/execution/TestOperator.scala]: How about mark these tests as ignore instead of deleting them?

**Comment by lgbo-ustc** [comment]
The coredump maybe related to one of latest pr in clickhouse backend.

**Comment by lgbo-ustc** [comment]
try to rebase to the latest main branch

File: shims/common/src/main/scala/io/glutenproject/GlutenConfig.scala [hunk 8]
```diff
@@ -87,7 +87,7 @@ class GlutenConfig(conf: SQLConf) extends Logging {
 
   // enable or disable columnar union
   val enableColumnarUnion: Boolean =
-    conf.getConfString("spark.gluten.sql.columnar.union", "true").toBoolean
+    conf.getConfString("spark.gluten.sql.columnar.union", "false").toBoolean
```
Comment (zzcclp) [inline shims/common/src/main/scala/io/glutenproject/GlutenConfig.scala:90]: why set to false as the default value ?

**Comment by zzcclp** [comment]
I ran the below sql and the Spark UI show the DAG below, there are r2c and c2r operators:
```
select count(date_sk) from (
  select d_date_sk as date_sk from date_dim
  union all
  select ws_sold_date_sk as date_sk from web_sales
)
```

image

**Comment by lgbo-ustc** [comment]
If don't wrap union exec transformer into a `WholeStageCodegenTransformer `, we need to implement `doExecuteColumnar`

**Comment by rui-mo** [comment]
Hi @lgbo-ustc, I'm testing on TPC-DS Q5, and found agg and union were in a single wholestage transformer. Is this the expected behavior?
Screenshot from 2022-10-28 08-06-14

**Comment by lgbo-ustc** [comment]
@zzcclp @rui-mo would you take a review again ?

**Review by rui-mo — APPROVED** [review]
Verified on TPC-DS. LGTM.

File: backends-clickhouse/src/test/scala/io/glutenproject/execution/GlutenClickHouseTPCDSParquetSuite.scala [hunk 9]
```diff
@@ -41,6 +41,7 @@ class GlutenClickHouseTPCDSParquetSuite extends GlutenClickHouseTPCDSAbstractSui
       // Currently, it can not support to read multiple partitioned file in one task.
       .set("spark.sql.files.maxPartitionBytes", "134217728")
       .set("spark.sql.files.openCostInBytes", "134217728")
+      .set("spark.gluten.sql.columnar.union", "true")
```
Comment (zzcclp) [inline backends-clickhouse/src/test/scala/io/glutenproject/execution/GlutenClickHouseTPCDSParquetSuite.scala:44]: remove this code

**Review by zzcclp — APPROVED** [review]
LGTM

**Pull request merged**